src/main/scala/scala2021/ashinkarev/task07/Utils.scala

```diff
@@ -0,0 +1,73 @@
+package scala2021.ashinkarev.task07
+
+import scala.util.{Try,Success,Failure}
+import java.io.File
+import java.io.FileInputStream
+
+trait Connection {
+  def close();
+
+  def run();
+}
+
+class PrintConnection(port: Int) extends Connection {
+  def close() = {
+    println("Closed")
+  }
+  def run() = {
+    println("Run")
+  }
+}
+
+object Utils {
+  def withConnection[TResult](port: Int)(body: (Connection) => TResult) = {
+    withResource(() => new PrintConnection(port))(body)((connection) => connection.close())
+  }
+
+  def withFile[TResult](path: String)(body: (File) => TResult) = {
+    withResource(() => new File(path))(body)()
+  }
+
+  def withFileInputStream[TResult](path: String)(body: (FileInputStream) => TResult) = {
+    withResource(() => new FileInputStream(path))(body)((fileInputStream: FileInputStream) => fileInputStream.close())
+  }
+
+  def withResource[TResource, TResult]
+    (resourceFactory: () => TResource)
+    (body: (TResource) => TResult)
+    (cleanup: (TResource) => Unit = (resource: TResource) => {}): TResult = {
+    var resource: TResource = null.asInstanceOf[TResource]
```

Comment:
It can be done without mutability and nulls. Looks very weird

Reply:
Try\catch is ok, you can use it for sure

Reply:
I needed to create a variable where I put resource if it was created successfully. It is the same pattern in .net using. How to do it correctly saving this feature for accessing resource?